stability: nil pointer panic in client.Txn.CleanupOnError

On the register cluster running beta-20160629, two nodes have failed (ten minutes apart) with the following panic:

```
E160717 06:28:28.001279 internal/client/txn.go:364  failure aborting transaction: does not exist; abort caused by: does not exist
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x140 pc=0xb685a0]

goroutine 3094076 [running]:
panic(0x18ba060, 0xc82000e100)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/internal/client.(*Txn).sendEndTxnReq(0x0, 0xc830067400, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:440 +0x50
github.com/cockroachdb/cockroach/internal/client.(*Txn).Rollback(0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:433 +0x3b
github.com/cockroachdb/cockroach/internal/client.(*Txn).CleanupOnError(0x0, 0x7f09906bda70, 0xc835e66ba0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:363 +0x92
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8201a3860, 0x7f099136cac8, 0xc823552780, 0xc82351d000, 0xc820a519cb, 0x27, 0x0, 0x0, 0x0, 0xc820a51900)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:502 +0xd07
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8201a3860, 0x7f099136cac8, 0xc823552780, 0xc82351d000, 0xc820a519cb, 0x27, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:361 +0xf6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc826b4c000, 0x7f099136cac8, 0xc823552780, 0xc820a519cb, 0x27, 0x0, 0x0, 0x0, 0x0, 0x7f09913b7201, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:640 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc826b4c000, 0x7f099136cac8, 0xc823552780, 0xc826b4c028, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:320 +0xe8
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc826b4c000, 0xc820d190e0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:275 +0x100c
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc820389a70, 0x7f09913711b0, 0xc8202de000, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:229 +0x98f
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x7f09913bc9b0, 0xc820584210)
        /go/src/github.com/cockroachdb/cockroach/server/server.go:370 +0x42
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc8200d02b0, 0x7f09913bc9b0, 0xc820584210, 0xc82034e010)
        /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:133 +0x333
```

This happened again on gamma, running beta-20160714.

```
W160718 20:45:07.289427 sql/lease.go:713  error releasing lease "51(\"blocks\") ver=1:1468509533172667968": does not exist
E160718 20:45:07.289708 internal/client/txn.go:371  failure aborting transaction: does not exist; abort caused by: does not exist
E160718 20:45:07.290614 internal/client/txn.go:371  failure aborting transaction: does not exist; abort caused by: does not exist
W160718 20:45:07.290666 sql/lease.go:713  error releasing lease "51(\"blocks\") ver=1:1468558242089988743": does not exist
E160718 20:45:07.291070 internal/client/txn.go:371  failure aborting transaction: does not exist; abort caused by: does not exist
W160718 20:45:07.291108 sql/lease.go:713  error releasing lease "55(\"comments\") ver=1:1468511843555426737": does not exist
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x140 pc=0xb7b680]

goroutine 7150662383 [running]:
panic(0x18e5680, 0xc820012080)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/internal/client.(*Txn).sendEndTxnReq(0x0, 0xc8c443e400, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:447 +0x50
github.com/cockroachdb/cockroach/internal/client.(*Txn).Rollback(0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:440 +0x3b
github.com/cockroachdb/cockroach/internal/client.(*Txn).CleanupOnError(0x0, 0x7f0340e3ff20, 0xc88d64f9f0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:370 +0x92
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8201e5b30, 0x7f08915fb1d0, 0xc8c2ec9840, 0xc8545d6000, 0xc85901c2b0, 0x56, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:505 +0xd07
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8201e5b30, 0x7f08915fb1d0, 0xc8c2ec9840, 0xc8545d6000, 0xc85901c2b0, 0x56, 0xc84c8a7688, 0x0, 0x0, 0x0, .
..)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:364 +0xf6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc89a03c300, 0x7f08915fb1d0, 0xc8c2ec9840, 0xc85901c2b0, 0x56, 0xc84c8a7688, 0x2a32498, 0x0, 0x0, 0x0,
 ...)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:640 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleExecute(0xc89a03c300, 0x7f08915fb1d0, 0xc8c2ec9840, 0xc89a03c328, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:628 +0x429
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc89a03c300, 0xc87c16e1c0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:298 +0xf03
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc8203e5980, 0x7f08916902f8, 0xc86986e900, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:229 +0x98f
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x7f08915be9b0, 0xc8c0162000)
        /go/src/github.com/cockroachdb/cockroach/server/server.go:370 +0x42
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc820158178, 0x7f08915be9b0, 0xc8c0162000, 0xc8200120b0)
        /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:133 +0x333
```

Additional logging from the just relocated beta cluster:
```
E160723 17:36:40.065839 sql/executor.go:499  txnState not cleared while txn == nil: &{txn:<nil> State:Aborted retrying:false retryIntent:false autoRetry:true commitSeen:false schemaChangers:{curGroupNum:0 curStatementIdx:0 schemaChangers:[]} tr:0xc82195acc0 sqlTimestamp:{sec:63604889966 nsec:625089321 loc:0x26fb5e0}}, execOpt {AutoRetry:true AutoCommit:true MinInitialTimestamp:1469293166.625087851,0}, stmts INSERT INTO blocks(block_id, writer_id, block_num, raw_bytes) VALUES ($1, $2, $3, $4), remaining
E160723 17:36:40.065945 sql/executor.go:510  AutoCommitError on nil txn: does not exist, txnState &{txn:<nil> State:Aborted retrying:false retryIntent:false autoRetry:true commitSeen:false schemaChangers:{curGroupNum:0 curStatementIdx:0 schemaChangers:[]} tr:0xc82195acc0 sqlTimestamp:{sec:63604889966 nsec:625089321 loc:0x26fb5e0}}, execOpt {AutoRetry:true AutoCommit:true MinInitialTimestamp:1469293166.625087851,0}, stmts INSERT INTO blocks(block_id, writer_id, block_num, raw_bytes) VALUES ($1, $2, $3, $4), remaining
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x140 pc=0xba05e0]

goroutine 9259823 [running]:
panic(0x190b7a0, 0xc82000e090)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Recover(0xc8203a2d80)
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:175 +0x70
panic(0x190b7a0, 0xc82000e090)
        /usr/local/go/src/runtime/panic.go:443 +0x4e9
github.com/cockroachdb/cockroach/internal/client.(*Txn).sendEndTxnReq(0x0, 0x7fad6985d300, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:449 +0x50
github.com/cockroachdb/cockroach/internal/client.(*Txn).Rollback(0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:442 +0x3b
github.com/cockroachdb/cockroach/internal/client.(*Txn).CleanupOnError(0x0, 0x7fad63ec27e0, 0xc8428e0630)
        /go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:372 +0x92
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc820330000, 0x7fad698e6fe8, 0xc8304609c0, 0xc827766800, 0xc838bd2cdf, 0x56, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:514 +0x12d6
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc820330000, 0x7fad698e6fe8, 0xc8304609c0, 0xc827766800, 0xc838bd2cdf, 0x56, 0xc8430a5680, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:364 +0xf6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc82144b980, 0x7fad698e6fe8, 0xc8304609c0, 0xc838bd2cdf, 0x56, 0xc8430a5680, 0x2a74488, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:640 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleExecute(0xc82144b980, 0x7fad698e6fe8, 0xc8304609c0, 0xc82144b9a8, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:628 +0x429
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82144b980, 0xc8246ee620, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:298 +0xf03
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc82029e5d0, 0x7fad698e7300, 0xc826daa000, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:229 +0x98f
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x7fad6985f920, 0xc838ffc000)
        /go/src/github.com/cockroachdb/cockroach/server/server.go:370 +0x42
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc8203a2d80, 0xc820178138, 0x7fad6985f920, 0xc838ffc000, 0xc82038e000)
        /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:132 +0x8d
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:134 +0x346
```

Can be found on: `cockroach@104.196.157.72`

--- a/sql/executor.go
+++ b/sql/executor.go
@@ -492,10 +492,10 @@ func (e *Executor) execRequest(ctx context.Context, session *Session, sql string
 		}
 		// This is where the magic happens - we ask db to run a KV txn and possibly retry it.
 		txn := txnState.txn // this might be nil if the txn was already aborted.
-		err := txnState.txn.Exec(execOpt, txnClosure)
+		err := txn.Exec(execOpt, txnClosure)
 
 		// Until #7881 fixed.
-		if txnState.State != NoTxn && txnState.txn == nil {
+		if txnState.State != NoTxn && (txn == nil || txnState.txn == nil) {
 			log.Errorf(context.TODO(), "txnState not cleared while txn == nil: %+v, execOpt %+v, stmts %+v, remaining %+v", txnState, execOpt, stmts, remainingStmts)
 		}
 
@@ -506,12 +506,12 @@ func (e *Executor) execRequest(ctx context.Context, session *Session, sql string
 			lastResult := &results[len(results)-1]
 			if aErr, ok := err.(*client.AutoCommitError); ok {
 				// Until #7881 fixed.
-				if txnState.txn == nil {
+				if txn == nil {
 					log.Errorf(context.TODO(), "AutoCommitError on nil txn: %+v, txnState %+v, execOpt %+v, stmts %+v, remaining %+v", err, txnState, execOpt, stmts, remainingStmts)
 				}
 				lastResult.Err = aErr
 				e.txnAbortCount.Inc(1)
-				txnState.txn.CleanupOnError(err)
+				txn.CleanupOnError(err)
 			}
 			if lastResult.Err == nil {
 				log.Fatalf(context.TODO(), "error (%s) was returned, but it was not set in the last result (%v)", err, lastResult)

